README.md
<p align="center">
  <h1 align="center">📡 Azure IoT 📍</h1>
  <p align="center">
    <i>Sending data from the edge</i>
  </p>
</p>

<p align="center">
  <a href="https://choosealicense.com/licenses/mit/">
    <img alt="license" src="https://img.shields.io/github/license/sjimenez44/AzureIot?style=flat-square&color=lightblue"/>
  </a>
  <a href="https://github.com/sjimenez44/AzureIot/actions/workflows/ci.yml/">
    <img alt="tests status" src="https://github.com/sjimenez44/AzureIot/actions/workflows/ci.yml/badge.svg"/>
  </a>
</p>

<p align="center">
  <p align="center">
    Browse the <a href="https://github.com/sjimenez44/AzureIot/releases">releases</a>.
  </p>
</p>

------------------------

## Features

- 🧪 &nbsp; Receiving data from the edge and formatting it
- 🏭 &nbsp; Save the data in Azure Table Storage

## API Reference

#### POST save data of sensor

```
  POST /api/SaveData
```

| Parameter  | Type     | Description                       |
| :--------- | :------- | :-------------------------------- |
| `sensorId` | `string` | **Required**. Id of the sensor    |
| `type`     | `string` | **Required**. Type of the data    |
| `value`    | `double` | **Required**. Value sensed        |

## Environment Variables

To run this project, you will need to add the following file `app.config` with the following content:

```json
{
  "useDevelopmentStorage": false,
  "connectionString": "<your_connection_string>",
  "tableName": "SensorData"
}
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-10-22
### Added
- Formatting data from sensors
- Save data to db

[Unreleased]: https://github.com/sjimenez44/AzureIot/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/sjimenez44/AzureIot/releases/tag/v1.0.0


tests/sensor_formatter.js
const chai = require('chai');
const expect = chai.expect;

const Sensor = require('../modules/data/Sensor');
const Temperature = require('../modules/data/Temperature');
const Humidity = require('../modules/data/Humidity');

const sensor = new Sensor();

describe("Data test module: ", function() {
  describe("Check sensor type strategies: ", function() {
    it("Check the returned value using: Temperature strategie ", function() {
      let result = sensor.formatter(new Temperature(25));
      expect(result).to.equal('77.000');
    });
    it("Check the returned value using: Humidity strategie ", function() {
      let result = sensor.formatter(new Humidity(2048));
      expect(result).to.equal('50.000');
    });
  });
});

tests/network_formatter.js
const chai = require('chai');
const expect = chai.expect;

const response = require('../network/response');

describe("Network test formatter: ", function() {
  describe("Check success formatter: ", function() {
    it("Check the returned value using: success without message", function() {
      let result = response.success(200, null);
      expect(result.status).to.equal(200);
      expect(result.body).to.equal('Ok');
    });
    it("Check the returned value using: success without status", function() {
      let result = response.success(null, 'Accepted');
      expect(result.status).to.equal(200);
      expect(result.body).to.equal('Accepted');
    });
    it("Check the returned value using: success with status and message", function() {
      let result = response.success(200, 'Accepted');
      expect(result.status).to.equal(200);
      expect(result.body).to.equal('Accepted');
    });
    it("Check the returned value using: success without status and message", function() {
      let result = response.success(null, null);
      expect(result.status).to.equal(200);
      expect(result.body).to.equal('Ok');
    });
  });
  describe("Check error formatter: ", function() {
    it("Check the returned value using: error without message and details", function() {
      let result = response.error(400, null, null);
      expect(result.status).to.equal(400);
      expect(result.body.error.message).to.equal('Invalid format');
      expect(result.body.error.details).to.equal(null);
    });
    it("Check the returned value using: error without status, message and details", function() {
      let result = response.error(null, null, null);
      expect(result.status).to.equal(500);
      expect(result.body.error.message).to.equal('Internal error');
      expect(result.body.error.details).to.equal(null);
    });
    it("Check the returned value using: error with details", function() {
      let result = response.error(400, null, 'Missing values');
      expect(result.status).to.equal(400);
      expect(result.body.error.message).to.equal('Invalid format');
      expect(result.body.error.details).to.equal('Missing values');
    });
  });
});

SaveData/index.js
const responsef = require('../network/response');
const process = require('./processdata');
const save = require('./savedata');

module.exports = async function (context, req) {
  if (req.body && req.body.type) {
    if (req.body.id) {
      if (req.body.value) {
        const sensor = process.processdata(req.body);
        const result = await save.savedata(sensor);
        (result)
          ? context.res = responsef.success(201, null)
          : context.res = responsef.error(400, null, null);
      } else {
        context.res = responsef.error(400, null, 
          'Please insert the data value');
      }
    } else {
      context.res = responsef.error(400, null, 
        'Please define the sensor id');
    }
  } else {
    context.res = responsef.error(400, null, 
      'Please define the type of the imput');
  }
}

SaveData/savedata.js
const guid = require('node-uuid');
const Storage = require('azure-storage');
const config = require('../config');

const entityGen = Storage.TableUtilities.entityGenerator;
const storageClient = Storage.createTableService(config.connectionString);
const tableName = config.tableName;

/**
* Creates a data entity to be used in library calls based on provided parameters
* @ignore
*
* @param {string}  sensorId   Identifier of the sensor
* @param {string}  type       Type of the sensor
* @param {string}  scale      Scale of measurement
* @param {string}  value      Value sensed by the sensor
* @return {Object}
*/
function createDataEntityDescriptor(sensorId, type, scale, value) {
  const dataEntity = {
    PartitionKey: entityGen.String(sensorId),
    RowKey: entityGen.String(guid.v4()),
    SensorId: entityGen.String(sensorId),
    Type: entityGen.String(type),
    Scale: entityGen.String(scale),
    Value: entityGen.Double(value)
  };
  return dataEntity;
}

exports.savedata = async function(data) {
  const dataEntity = createDataEntityDescriptor(data.id, data.type, data.scale, data.value);
  storageClient.insertEntity(tableName, dataEntity, (error, result, response) => {
    if (error) {
      return false;
    }
  });
  return true;
};

network/response.js
//-- Diccionario de mensajes segun codigo de estado
const statusMessages = {
    '200': 'Ok',
    '201': 'Created',
    '202': 'Accepted',
    '204': 'No content',
    '301': 'Moved Permanently',
    '400': 'Invalid format',
    '401': 'Unauthorized',
    '402': 'Payment required',
    '403': 'Forbidden',
    '404': 'Not found',
    '406': 'Not acceptable',
    '409': 'Conflict',
    '429': 'Too many requests',
    '500': 'Internal error',
    '520': 'Operation failed'
  }
  
  //-- Realiza un formato para las respuestas satisfactorias
  exports.success = function (status, message) {
    let statusCode = status;
    let statusMessage = message;
    
    //-- Si el estado no esta definido toma por defecto 200
    if (!status) {
      statusCode = 200;
    }
  
    /* 
      Si el mensaje no esta definido por defecto toma el
      el valor del diccionario
    */
    if (!message) {
      statusMessage = statusMessages[statusCode];
    }
  
    return {
      status: statusCode,
      body: statusMessage,
    }
  }
  
  //-- Realiza un formato para las respuestas erroneas
  exports.error = function (status, message, details) {
    let statusCode = status;
    let statusMessage = message;
    let statusDetails = details;
    
    //-- Si el estado no esta definido toma por defecto 500
    if (!status) {
      statusCode = 500;
    }
  
    /* 
      Si el mensaje no esta definido por defecto toma el
      el valor del diccionario
    */
    if (!message) {
      statusMessage = statusMessages[statusCode];
    }
  
    return {
      status: statusCode,
      body: {
        error: {
          message: statusMessage,
          details: statusDetails
        }
      }
    }
  }